src/txrequest.cpp

```diff
@@ -0,0 +1,238 @@
+        // ByTxid index.
+        if (it_next != m_index.get<ByTxid>().end() && it_next->m_txid == it->m_txid &&
+            it_next->GetState() == State::CANDIDATE_READY) {
+            // If one such CANDIDATE_READY exists (for this txid), convert it to CANDIDATE_BEST.
+            Modify<ByTxid>(it_next, [](Entry& entry){ entry.SetState(State::CANDIDATE_BEST); });
+        }
+    }
+    Modify<ByTxid>(it, [new_state](Entry& entry){ entry.SetState(new_state); });
+    assert(!it->IsSelected());
+}
+
+bool TxRequestTracker::MakeCompleted(typename TxRequestTracker::Index::index<ByTxid>::type::iterator it)
+{
+    // Nothing to be done if it's already COMPLETED.
+    if (it->GetState() == State::COMPLETED) return true;
+
+    if ((it == m_index.get<ByTxid>().begin() || std::prev(it)->m_txid != it->m_txid) &&
+        (std::next(it) == m_index.get<ByTxid>().end() || std::next(it)->m_txid != it->m_txid ||
+        std::next(it)->GetState() == State::COMPLETED)) {
+        // This is the first entry for this txid, and the last non-COMPLETED one. There are only COMPLETED ones left.
+        // Delete them all.
+        uint256 txid = it->m_txid;
+        do {
+            it = Erase<ByTxid>(it);
+        } while (it != m_index.get<ByTxid>().end() && it->m_txid == txid);
+        return false;
+    }
+
+    // Mark the entry COMPLETED, and select the best best entry if needed.
+    ChangeAndReselect(it, State::COMPLETED);
+
+    return true;
+}
+
+void TxRequestTracker::SetTimePoint(std::chrono::microseconds now)
+{
+    // Iterate over all CANDIDATE_DELAYED and REQUESTED from old to new, as long as they're in the past,
+    // and convert them to CANDIDATE_READY and COMPLETED respectively.
+    while (!m_index.empty()) {
+        auto it = m_index.get<ByTime>().begin();
+        if (it->GetState() == State::CANDIDATE_DELAYED && it->m_time <= now) {
+            PromoteCandidateNew(m_index.project<ByTxid>(it));
+        } else if (it->GetState() == State::REQUESTED && it->m_time <= now) {
+            MakeCompleted(m_index.project<ByTxid>(it));
+        } else {
+            break;
+        }
+    }
+
+    while (!m_index.empty()) {
+        // If time went backwards, we may need to demote CANDIDATE_BEST and CANDIDATE_READY entries back
+        // to CANDIDATE_DELAYED. This is an unusual edge case, and unlikely to matter in production. However,
+        // it makes it much easier to specify and test TxRequestTracker's behaviour.
+        auto it = std::prev(m_index.get<ByTime>().end());
+        if (it->IsSelectable() && it->m_time > now) {
+            ChangeAndReselect(m_index.project<ByTxid>(it), State::CANDIDATE_DELAYED);
+        } else {
+            break;
+        }
+    }
+}
+
+void TxRequestTracker::AlreadyHaveTx(const uint256& txid)
+{
+    auto it = m_index.get<ByTxid>().lower_bound(EntryTxid{txid, State::CANDIDATE_DELAYED, 0});
```

Comment:
259fb92

Maybe add as a comment "If transaction has already been evaluated once for this block tip, delete unconditionally all its entries."

Is this a slight behavior change ?  Previously we would have only delete transaction from download for this peer. If you received a block between peers processing you have flushed `recentRejects` and so here return a different evaluation.

Reply:
> Maybe add as a comment "If transaction has already been evaluated once for this block tip, delete unconditionally all its entries."

I don't understand what reevaluation has to do with that. And `TxRequestTracker` doesn't know or care about the concept of blocks. It's just told that a transaction is no longer needed, so it can forget announcements about it. Why the caller decided that was the right thing to do is not its concern.

> Is this a slight behavior change ?

The whole thing is a giant behavior change - it's changing out the semantics for something conceptually similar, but very different in the details.

Reply:
> Why the caller decided that was the right thing to do is not its concern.

Right, I understand comments should be concerned with other transaction relay mechanisms.